api/src/main/java/au/org/aodn/nrmn/restapi/service/upload/S3IO.java

```diff
@@ -1,41 +1,91 @@
 package au.org.aodn.nrmn.restapi.service.upload;
 
 import java.io.ByteArrayOutputStream;
-import java.io.InputStream;
+import java.io.File;
+import java.security.SecureRandom;
+import java.time.LocalDateTime;
+import java.time.OffsetDateTime;
+import java.util.List;
 
+import org.apache.commons.lang3.RandomStringUtils;
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.stereotype.Service;
 import org.springframework.web.multipart.MultipartFile;
 
 import software.amazon.awssdk.awscore.exception.AwsServiceException;
 import software.amazon.awssdk.core.sync.RequestBody;
 import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.CopyObjectRequest;
+import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
+import software.amazon.awssdk.services.s3.model.ObjectCannedACL;
 import software.amazon.awssdk.services.s3.model.PutObjectRequest;
-import software.amazon.awssdk.services.s3.model.PutObjectResponse;
 
 @Service
 public class S3IO {
 
     @Value("${app.s3.bucket}")
     private String bucket;
 
+    @Value("${app.s3.bucket-shared}")
+    private String bucketShared;
+
     S3Client client;
 
     public S3Client getClient() {
-        if(client == null) client = S3Client.create();
+        if (client == null)
+            client = S3Client.create();
         return client;
     }
 
+    public void uploadEndpoint(String viewName, File file) {
+        try {
+            var fullPath = String.join("/", List.of("endpoints", viewName + ".csv"));
+            var res = RequestBody.fromFile(file);
+            getClient().putObject(PutObjectRequest.builder().bucket(bucket).key(fullPath).build(), res);
+        } catch (AwsServiceException e) {
+            throw new RuntimeException("Failed to write to S3: " + e.getMessage());
+        }
+    }
+
+    public String createS3Link(String viewName, LocalDateTime expires) throws Exception {
+        try {
+            var sessionId = RandomStringUtils.random(20, 0, 0, true, true, null, new SecureRandom());
+            var sourceKey = String.join("/", List.of("endpoints", viewName + ".csv"));
+            var destinationKey = String.join("/", List.of("endpoints", sessionId, viewName + ".csv"));
+            var request = CopyObjectRequest.builder()
+                    .sourceBucket(bucket).sourceKey(sourceKey)
+                    .destinationBucket(bucketShared).destinationKey(destinationKey)
+                    .acl(ObjectCannedACL.PUBLIC_READ)
+                    .expires(expires.toInstant(OffsetDateTime.now().getOffset()))
+                    .build();
+            client.copyObject(request);
```

Comment:
Bug here should call getClient() instead of client otherwise null pointer exception for client, alternatively use @PostConstruct to init the client object is fine too

Reply:
Bug: Create the share link with past expires date (for example today is 4 Jan, you create a link of 3 Jan) then copy the link and try to download it still works

Reply:
Bug: Click the (x) in the expires field, it become dd/mm/yyyy, enter other info and click the create link will cause internal server error, I think it is better to either block user from click the button or give a better message instead of 500